src/core/border_router/routing_manager.cpp

```diff
@@ -1350,116 +1173,418 @@ bool RoutingManager::UpdateRouterAdvertHeader(const Ip6::Nd::RouterAdvertMessage
+{
+    Ip6::Prefix prefix;
+    Entry *     entry;
+
+    VerifyOrExit(aPio.IsValid());
+    aPio.GetPrefix(prefix);
+
+    VerifyOrExit(Get<RoutingManager>().ShouldProcessPrefixInfoOption(aPio, prefix));
+
+    LogInfo("Processing PIO (%s, %u seconds)", prefix.ToString().AsCString(), aPio.GetValidLifetime());
+
+    entry = mEntries.FindMatching(Entry::Matcher(prefix, Entry::kTypeOnLink));
+
+    if (entry == nullptr)
+    {
+        VerifyOrExit(aPio.GetValidLifetime() != 0);
+
+        if (mEntries.IsFull())
         {
-            // Check for most recent stale OMR Prefixes
-            nextStaleTime = OT_MIN(nextStaleTime, prefixStaleTime);
+            LogWarn("Discovered too many prefixes, ignore on-link prefix %s", prefix.ToString().AsCString());
+            ExitNow();
         }
+
+        SuccessOrExit(Get<RoutingManager>().PublishExternalRoute(prefix, NetworkData::kRoutePreferenceMedium));
+        entry = mEntries.PushBack();
+        entry->InitFrom(aPio);
+    }
+    else
+    {
+        Entry newEntry;
+
+        newEntry.InitFrom(aPio);
+        entry->AdoptValidAndPreferredLiftimesFrom(newEntry);
     }
 
-    if (requireCheckStaleOnlinkPrefix)
+    mTimer.FireAtIfEarlier(entry->GetExpireTime());
+    SignalTableChanged();
+
+exit:
+    return;
+}
+
+void RoutingManager::DiscoveredPrefixTable::ProcessRouteInfoOption(const Ip6::Nd::RouteInfoOption &aRio)
+{
+    Ip6::Prefix prefix;
+    Entry *     entry;
+
+    VerifyOrExit(aRio.IsValid());
+    aRio.GetPrefix(prefix);
+
+    VerifyOrExit(Get<RoutingManager>().ShouldProcessRouteInfoOption(aRio, prefix));
+
+    LogInfo("Processing RIO (%s, %u seconds)", prefix.ToString().AsCString(), aRio.GetRouteLifetime());
+
+    entry = mEntries.FindMatching(Entry::Matcher(prefix, Entry::kTypeRoute));
+
+    if (aRio.GetRouteLifetime() == 0)
     {
-        nextStaleTime = OT_MIN(nextStaleTime, maxOnlinkPrefixStaleTime);
+        VerifyOrExit(entry != nullptr);
+
+        Get<RoutingManager>().UnpublishExternalRoute(entry->GetPrefix());
+        mEntries.Remove(*entry);
+
+        ExitNow();
     }
 
-    if (nextStaleTime == now.GetDistantFuture())
+    if (entry == nullptr)
     {
-        if (mDiscoveredPrefixStaleTimer.IsRunning())
+        if (mEntries.IsFull())
         {
-            LogDebg("Prefix stale timer stopped");
+            LogWarn("Discovered too many prefixes, ignore new prefix %s", prefix.ToString().AsCString());
+            ExitNow();
         }
-        mDiscoveredPrefixStaleTimer.Stop();
+
+        SuccessOrExit(Get<RoutingManager>().PublishExternalRoute(prefix, aRio.GetPreference()));
+        entry = mEntries.PushBack();
     }
-    else
+
+    entry->InitFrom(aRio);
+
+    mTimer.FireAtIfEarlier(entry->GetExpireTime());
+    SignalTableChanged();
+
+exit:
+    return;
+}
+
+void RoutingManager::DiscoveredPrefixTable::FindFavoredOnLinkPrefix(Ip6::Prefix &aPrefix) const
+{
+    // Find the smallest (not deprecated) on-link prefix entry in the
```

Comment:
nit: RFC 4861 term for non-deprecated is preferred.
```suggestion
    // Find the smallest (preferred) on-link prefix entry in the
```